rust/src/bittorrent_dht/bittorrent_dht.rs

```diff
@@ -0,0 +1,401 @@
+    tx: *mut std::os::raw::c_void, _direction: u8,
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, BitTorrentDHTTransaction);
+
+    // Transaction is done if we have a request, response, or error since
+    // a new transaction is created for each received packet
+    if tx.request.is_some() || tx.response.is_some() || tx.error.is_some() {
+        return 1;
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_bittorrent_dht_state_get_events(
+    tx: *mut std::os::raw::c_void,
+) -> *mut core::AppLayerDecoderEvents {
+    let tx = cast_pointer!(tx, BitTorrentDHTTransaction);
+    return tx.events;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_bittorrent_dht_state_get_event_info(
+    event_name: *const std::os::raw::c_char, event_id: *mut std::os::raw::c_int,
+    event_type: *mut core::AppLayerEventType,
+) -> std::os::raw::c_int {
+    if event_name == std::ptr::null() {
+        return -1;
+    }
+
+    let event_name = unsafe { std::ffi::CStr::from_ptr(event_name) };
+    if let Ok(event_name) = event_name.to_str() {
+        if let Ok(event) = BitTorrentDHTEvent::from_str(event_name) {
+            unsafe {
+                *event_id = event as std::os::raw::c_int;
+                *event_type = core::APP_LAYER_EVENT_TYPE_TRANSACTION;
+            }
+        } else {
+            // Unknown event...
+            return -1;
+        }
+    } else {
+        // UTF-8 conversion failed. Should not happen.
+        return -1;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_bittorrent_dht_state_get_event_info_by_id(
+    event_id: std::os::raw::c_int, event_name: *mut *const std::os::raw::c_char,
+    event_type: *mut core::AppLayerEventType,
+) -> i8 {
+    if let Some(e) = BitTorrentDHTEvent::from_id(event_id as u32) {
+        unsafe {
+            *event_name = e.to_cstring().as_ptr() as *const std::os::raw::c_char;
+            *event_type = core::APP_LAYER_EVENT_TYPE_TRANSACTION;
+        }
+        return 0;
+    }
+    return -1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_bittorrent_dht_state_get_tx_iterator(
+    _ipproto: u8, _alproto: AppProto, state: *mut std::os::raw::c_void, min_tx_id: u64,
+    _max_tx_id: u64, istate: &mut u64,
+) -> applayer::AppLayerGetTxIterTuple {
+    let state = cast_pointer!(state, BitTorrentDHTState);
+    match state.tx_iterator(min_tx_id, istate) {
+        Some((tx, out_tx_id, has_next)) => {
+            let c_tx = unsafe { transmute(tx) };
+            let ires = applayer::AppLayerGetTxIterTuple::with_values(c_tx, out_tx_id, has_next);
+            return ires;
+        }
+        None => {
+            return applayer::AppLayerGetTxIterTuple::not_found();
+        }
+    }
+}
+
+// Parser name as a C style string.
+const PARSER_NAME: &[u8] = b"bittorrent-dht\0";
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_bittorrent_dht_udp_register_parser() {
+    let default_port = CString::new("[1024:65535]").unwrap();
```

Comment:
I wonder if this large port range somehow is related to the CI fuzz timeout warning. Do you notice any slow down in startup with this?

Reply:
Hmm that's a good question but I'm not sure how to measure the startup slowdown.